src/Docker.DotNet/Endpoints/StreamUtil.cs

```diff
@@ -10,37 +10,44 @@ namespace Docker.DotNet.Models
 {
     internal static class StreamUtil
     {
-        private static Newtonsoft.Json.JsonSerializer _serializer = new Newtonsoft.Json.JsonSerializer();
+        private static readonly Newtonsoft.Json.JsonSerializer _serializer = new Newtonsoft.Json.JsonSerializer();
 
         internal static async Task MonitorStreamAsync(Task<Stream> streamTask, DockerClient client, CancellationToken cancel, IProgress<string> progress)
         {
+            var tcs = new TaskCompletionSource<string>();
+            cancel.Register(() => tcs.TrySetCanceled(default));
+
             using (var stream = await streamTask)
             {
-                // ReadLineAsync must be cancelled by closing the whole stream.
-                using (cancel.Register(() => stream.Dispose()))
+                using (var reader = new StreamReader(stream, new UTF8Encoding(false)))
                 {
-                    using (var reader = new StreamReader(stream, new UTF8Encoding(false)))
+                    while (!cancel.IsCancellationRequested)
                     {
-                        string line;
-                        while ((line = await reader.ReadLineAsync()) != null)
+                        string line = await await Task.WhenAny(reader.ReadLineAsync(), tcs.Task);

```

Comment:
Fascinating idea. You seem to have an extra `await` though.

Reply:
Yes, it looks like that.

`await Task.WhenAny(task1, task2)` would return either task1 or task2.
Then, this task1 or task2 need to be awaited too.

Original idea from #343 by @suneption, I just adjusted it for tests to pass.

Reply:
Crazy. I did not know you could inline await :).